metals/src/main/scala/scala/meta/internal/metals/FileWatcher.scala

```diff
@@ -37,62 +37,103 @@ final class FileWatcher(
-    def watch(dir: AbsolutePath, isSourceDirectory: Boolean): Unit = {
-      if (!dir.isDirectory) {
-        dir.createDirectories()
-        if (isSourceDirectory) createdSourceDirectories.add(dir)
+    def watch(path: AbsolutePath, isSource: Boolean): Unit = {
+      if (!path.isDirectory && !path.isFile) {
+        val pathToCreate = if (path.isScalaOrJava) {
+          AbsolutePath(path.toNIO.getParent())
+        } else {
+          path
+        }
+        pathToCreate.createDirectories()
+        if (isSource) createdSourceDirectories.add(pathToCreate)
       }
-      directoriesToWatch.add(dir.toNIO)
+      if (path.isScalaOrJava) sourcesFilesToWatch.add(path.toNIO)
+      else sourcesDirectoriesToWatch.add(path.toNIO)
     }
     // Watch the source directories for "goto definition" index.
-    buildTargets.sourceDirectories.foreach(watch(_, isSourceDirectory = true))
+    buildTargets.sourceItems.foreach(watch(_, isSource = true))
     buildTargets.scalacOptions.foreach { item =>
       // Watch META-INF/semanticdb directories for "find references" index.
       watch(
         item.targetroot.resolve(Directories.semanticdb),
-        isSourceDirectory = false
+        isSource = false
       )
     }
-    startWatching(directoriesToWatch)
+    startWatching(sourcesFilesToWatch, sourcesDirectoriesToWatch)
     createdSourceDirectories.asScala.foreach(_.delete())
   }
 
-  private def startWatching(paths: util.List[Path]): Unit = {
+  private def startWatching(
+      files: util.List[Path],
+      directories: util.List[Path]
+  ): Unit = {
     stopWatching()
-    val watcher = DirectoryWatcher
+    val directoryWatcher = DirectoryWatcher
+      .builder()
+      .paths(directories)
+      .listener(new DirectoryListener())
+      .build()
+    activeDirectoryWatcher = Some(directoryWatcher)
+    directoryWatching = directoryWatcher.watchAsync(directoryExecutor)
+
+    val fileWatcher = DirectoryWatcher
       .builder()
-      .paths(paths)
-      .listener(new Listener)
+      .paths(files.map(_.getParent()))
+      .listener(new FileListener(files = files.asScala.toSet))
       .build()
-    activeWatcher = Some(watcher)
-    watching = watcher.watchAsync(executor)
+    activeFileWatcher = Some(fileWatcher)
+    fileWatching = fileWatcher.watchAsync(fileExecutor)
   }
 
   private def stopWatching(): Unit = {
-    activeWatcher.foreach(_.close())
-    watching.cancel(false)
+    activeDirectoryWatcher.foreach(_.close())
+    activeFileWatcher.foreach(_.close())
+    fileWatching.cancel(false)
+    directoryWatching.cancel(false)
   }
 
-  class Listener extends DirectoryChangeListener {
+  class DirectoryListener extends DirectoryChangeListener {
     override def onEvent(event: DirectoryChangeEvent): Unit = {
       if (!Files.isDirectory(event.path())) {
         didChangeWatchedFiles(event)
       }
     }
   }
 
+  class FileListener(files: Set[Path]) extends DirectoryChangeListener {
```

Comment:
```suggestion
  class FileListener(watched: Set[Path]) extends DirectoryChangeListener {
```
this way, we can have `(...) && watched(event.path())` instead of `(...) && files(event.path())`